internal/worker/worker.go

```diff
@@ -0,0 +1,444 @@
+	)
+	defer span.End()
+
+	task, err := newTaskFromCondition(condition)
+	if err != nil {
+		w.logger.WithError(err).Warn("error initializing task from condition")
+
+		w.eventAckComplete(e)
+
+		metrics.RegisterEventCounter(false, "ack")
+		metrics.RegisterSpanEvent(span, condition, w.id.String(), "", "sent ack, error task init")
+
+		return
+	}
+
+	startTS := time.Now()
+
+	publisher, err := newStatusKVPublisher(w.stream, w.logger, w.id.String(), w.cfg.NatsOptions.KV)
+	if err != nil {
+		w.logger.WithError(err).Warn("status KV init - internal error")
+
+		w.eventNak(e)
+
+		metrics.RegisterEventCounter(false, "nack")
+		metrics.RegisterSpanEvent(span, condition, w.id.String(), "", "sent nack, error task init")
+	}
+
+	// update task state, status
+	task.Status = "Collecting inventory outofband for device"
+	task.SetState(rctypes.Active)
+
+	w.logger.WithFields(logrus.Fields{
+		"deviceID":    task.Parameters.AssetID.String(),
+		"conditionID": task.ID,
+	}).Info(task.Status)
+
+	// publish update
+	publisher.Publish(ctx, task)
+
+	// check no error
+	err = w.runTaskWithMonitor(ctx, task, e)
+	switch err {
+	case errInventoryQuery:
+		// inventory lookup failure - non 404 errors
+		task.SetState(rctypes.Failed)
+		task.Status = err.Error()
+
+		metrics.RegisterEventCounter(true, "nack")
+		w.eventNak(e) // have the message bus re-deliver the message
+
+		metrics.RegisterSpanEvent(
+			span,
+			condition,
+			w.id.String(),
+			task.Parameters.AssetID.String(),
+			"sent nack: store query error",
+		)
+
+	case errAssetNotFound, errCollector:
+		// asset was not found
+		task.SetState(rctypes.Failed)
+		task.Status = err.Error()
+
+		w.eventAckComplete(e)
+
+		metrics.RegisterSpanEvent(
+			span,
+			condition,
+			w.id.String(),
+			task.Parameters.AssetID.String(),
+			"sent ack: error"+err.Error(),
+		)
+
+	case nil:
+		// work completed successfully
+		task.SetState(rctypes.Succeeded)
+		publisher.Publish(ctx, task)
+
+		w.eventAckComplete(e)
+
+		metrics.RegisterConditionMetrics(startTS, string(rctypes.Succeeded))
+		metrics.RegisterEventCounter(true, "ack")
+		metrics.RegisterSpanEvent(
+			span,
+			condition,
+			w.id.String(),
+			task.Parameters.AssetID.String(),
+			"sent ack: condition finalized",
+		)
+
+		publisher.Publish(ctx, task)
+		w.logger.WithFields(logrus.Fields{
+			"deviceID":    task.Parameters.AssetID.String(),
+			"conditionID": task.ID,
+			"elapsed":     time.Since(startTS).String(),
+			"state":       task.state,
+			"status":      task.Status,
+		}).Info("task for device completed")
+	}
+}
+
+// runTaskWithMonitor runs the task method based on the parameters, while ack'ing its progress to the NATS JS.
+func (w *Worker) runTaskWithMonitor(ctx context.Context, task *Task, e events.Message) error {
+	ctx, span := otel.Tracer(pkgName).Start(
+		ctx,
+		"worker.runTaskWithMonitor",
+	)
+	defer span.End()
+
+	// the child function is expected to close this channel to indicate its done
+	doneCh := make(chan bool)
```

Comment:
Consider making this channel `chan struct{}`? No one is reading this for a value.

_Edit:_ Also consider a comment here pointing out where the channel is closed.